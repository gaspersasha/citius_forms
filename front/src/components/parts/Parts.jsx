import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { getTestFlag } from '~utils';
import { useGOContext, useVehicleContext } from '~contexts';
import s from './parts.module.sass';

function validate(curr, max) {
  if (!curr || curr < 1) return 1;
  if (curr > max) return max;

  return curr;
}

export default function Parts({ part = 1, step = 1 }) {
  const { vehicle } = useVehicleContext();
  // GO a/b start
  // Display <Part /> component only within GO test "Serf Service Journey"
  const { isGoRdy } = useGOContext();
  const [variant, setVariant] = useState('0');

  useEffect(() => {
    if (!isGoRdy) return;
    setVariant(getTestFlag('ssj'));
  }, [isGoRdy]);

  if (variant !== '1' && !vehicle.leadgen_B) return null;
  // GO a/b end

  const totalPartsCount = 3;
  const partSubtitles = [
    [
      'Personal details',
      'Address details',
      'Employment details',
      'Bank details',
    ],
    ['Part exchange', 'Part exchange', 'Part exchange', 'Finance term'],
    ['Submit  application'],
  ];

  const validPart = validate(part, totalPartsCount);
  const validStep = validate(step, partSubtitles[validPart - 1].length);

  return (
    <section className={s.container}>
      <b>
        Part {validPart} of {totalPartsCount}
      </b>
      &nbsp;
      {partSubtitles[validPart - 1][validStep - 1]}
    </section>
  );
}

Parts.propTypes = {
  part: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
};
