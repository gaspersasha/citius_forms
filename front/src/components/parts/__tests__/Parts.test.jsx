import React from 'react';
import { useGOContext, useVehicleContext } from '~contexts';
import Parts from '../Parts';

jest.mock('~contexts', () => ({
  useGOContext: jest.fn(),
  useVehicleContext: jest.fn(),
}));

useGOContext.mockImplementation(() => ({ isGoRdy: true }));
useVehicleContext.mockImplementation(() => ({ vehicle: { leadgen_B: false } }));

describe('<Parts />', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'google_optimize', {
      value: {
        get: () => '1',
      },
      configurable: true,
      writable: true,
    });
  });

  it('no part props', () => {
    const wrapper = mount(<Parts step={5} part={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('no step props', () => {
    const wrapper = mount(<Parts part={2} step={0} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('part less than 1 props', () => {
    const wrapper = mount(<Parts part={-1} step={3} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('part bigger than totalPartsCount const props', () => {
    const wrapper = mount(<Parts part={8} step={3} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('step less than 1 props', () => {
    const wrapper = mount(<Parts part={1} step={0} />);

    expect(wrapper).toMatchSnapshot();
  });
});
