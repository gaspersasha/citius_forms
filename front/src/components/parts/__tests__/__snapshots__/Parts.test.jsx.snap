// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Parts /> no part props 1`] = `
<Parts
  part={0}
  step={5}
>
  <section
    className="container"
  >
    <b>
      Part 
      1
       of 
      3
    </b>
     
    Bank details
  </section>
</Parts>
`;

exports[`<Parts /> no step props 1`] = `
<Parts
  part={2}
  step={0}
>
  <section
    className="container"
  >
    <b>
      Part 
      2
       of 
      3
    </b>
     
    Part exchange
  </section>
</Parts>
`;

exports[`<Parts /> part bigger than totalPartsCount const props 1`] = `
<Parts
  part={8}
  step={3}
>
  <section
    className="container"
  >
    <b>
      Part 
      3
       of 
      3
    </b>
     
    Submit  application
  </section>
</Parts>
`;

exports[`<Parts /> part less than 1 props 1`] = `
<Parts
  part={-1}
  step={3}
>
  <section
    className="container"
  >
    <b>
      Part 
      1
       of 
      3
    </b>
     
    Employment details
  </section>
</Parts>
`;

exports[`<Parts /> step less than 1 props 1`] = `
<Parts
  part={1}
  step={0}
>
  <section
    className="container"
  >
    <b>
      Part 
      1
       of 
      3
    </b>
     
    Personal details
  </section>
</Parts>
`;
